finagle-http/src/test/scala/com/twitter/finagle/http/ServerSideDecodingTest.scala

```diff
@@ -0,0 +1,94 @@
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
+
+/** Provides tests for server side content decoding.
+  *
+  * If and when client side compression is implemented, this test should probably be removed in favour of a complete
+  * entry in [[EndToEndTest]]. Client side compression is currently made problematic by netty
+  * (see https://github.com/netty/netty/issues/4970).
+  */
+@RunWith(classOf[JUnitRunner])
+class ServerSideDecodingTest extends FunSuite with GeneratorDrivenPropertyChecks with BeforeAndAfterAll {
+  // Echo server (with decoding)
+  val server = finagle.Http.server
+    .withLabel("server")
+    .withDecompression(true)
+    .serve("localhost:*", new Service[Request, Response] {
+      def apply(request: Request) = {
+        val response = Response()
+        response.contentString = request.contentString
+        Future.value(response)
+      }
+    })
+
+  // Standard client
+  val client = ClientBuilder()
+    .codec(Http())
+    .hosts(Seq(server.boundAddress.asInstanceOf[InetSocketAddress]))
+    .hostConnectionLimit(1)
+    .build()
+
+  // URL at which all requests should be made
+  val url: String = {
+    val adr = server.boundAddress.asInstanceOf[InetSocketAddress]
+    s"http://${adr.getHostName}:${adr.getPort}/"
+  }
+
+  // Makes sure everything is cleaned up
+  override def afterAll(): Unit = {
+    Closable.all(client, server)
+  }
+
+  // Helper class - might be overkill to have a sum type for just one test, but it makes it simple to provide an
+  // Arbitrary instance for encoders and to make the actual test that much more readable.
+  sealed abstract class Encoder(val name: String) {
+    def encodeWith(out: OutputStream): OutputStream
+    def encode(string: String): Buf = {
+      val bytes = new ByteArrayOutputStream()
+      val out = new PrintStream(encodeWith(bytes), true, "UTF-8")
+      out.print(string)
+      out.close() // Do not remove, filter streams absolutely need this to generate legal content.
+      Buf.ByteArray.Shared(bytes.toByteArray)
+    }
+  }
+
+  case object Gzip extends Encoder(HttpHeaders.Values.GZIP) {
+    override def encodeWith(out: OutputStream) = new GZIPOutputStream(out)
+  }
+
+  case object Deflate extends Encoder(HttpHeaders.Values.DEFLATE) {
+    override def encodeWith(out: OutputStream) = new DeflaterOutputStream(out)
+  }
+
+  case object Identity extends Encoder(HttpHeaders.Values.IDENTITY) {
+    override def encodeWith(out: OutputStream) = out
+  }
+
+  implicit val arbEncoder: Arbitrary[Encoder] = Arbitrary(Gen.oneOf(Gzip, Deflate, Identity))
+
+  test("decode client-side encoded entity bodies") {
+    forAll { (content: String, encoder: Encoder) ⇒
+      val req = RequestBuilder()
+        .setHeader("Content-Encoding", encoder.name)
+        .setHeader("Content-Type", "text/plain;charset=utf-8")
+        .url(url)
+
+      assert(content == Await.result(client(req.buildPost(encoder.encode(content)))).contentString)
```

Comment:
This makes sense, but I feel I must point out I failed to do so because `EndToEndTest` mostly does not specify timeouts - I assumed this was the way things were done in Finagle.

By which I do not mean to say I won't fix this problem - I will, you're obviously right - but rather that `EndToEndTest` might need some work.